idempotence: sort vars before merging

vladimir-mencl-eresearch opened the pull request:
Python dicts are unordered.  And we get the list of variables to merge from a dict - so in arbitrary order.

That can have effect on the merged variables - and can produce spurious changes
on subsequent runs of the same playbook with the same input variables.
(E.g., when merging a dict of lists, the entries in the inner lists could be ordered
in different ways depending on the order in which the variables were merged)

This can get worked around by playbooks using this module by sorting the lists
before using them ... but that is sub-optimal.

Improve the situation by sorting the list of variable names before merging.

And, in recursive merging, also sort the keys of any dict being merged.

vladimir-mencl-eresearch committed:
idempotence: sort vars before merging

Python dicts are unordered.  And we get the list of variables to merge from a dict - so in arbitrary order.

That can have effect on the merged variables - and can produce spurious changes
on subsequent runs of the same playbook with the same input variables.
(E.g., when merging a dict of lists, the entries in the inner lists could be ordered
in different ways depending on the order in which the variables were merged)

This can get worked around by playbooks using this module by sorting the lists
before using them ... but that is sub-optimal.

Improve the situation by sorting the list of variable names before merging.

And, in recursive merging, also sort the keys of any dict being merged.

fajpunk commented:
Thanks!  I like the idea of sorting the keys used to generate `merge_vals` up front.  What do we get by sorting the keys of dict values that we want to merge though, given that the dict values that we want to merge are already in ordered deterministically by doing the first thing?

vladimir-mencl-eresearch commented:
Fair call.  I was thinking that each call to `dict.keys()` introduces non-determinism.
It would hold in the generic case (non-determinism of the execution), but I now agree that the determinism of the result is already achieved by sorting the values to merge.

Non-determinism in the execution order will not impact the returned value - does not matter in which order the dict entries of the result are populated; they will be anyway non-deterministic when iterated over.

Reverting the second change now...

vladimir-mencl-eresearch committed:
revert part of idempotence: sort vars - sorting vals is enough

No need to sort dict keys in recursive dict merge - deterministic result is
already achieved by sorting the values to merge.

This would only give deterministic *execution order*, but has no impact on the
returned value, so removing the second sort...

fajpunk commented:
Right on.  I'll merge this, update the docs, and cut a release for pypi.  May be tomorrow given that the tests take a billion years to run for all the python/ansible version combinations :grimacing:

Thanks again!

vladimir-mencl-eresearch commented:
Happy to help - we are making good use of the plugin, so I'm happy to contribute where we see it can do an even better job.

Cheers,
Vlad

vladimir-mencl-eresearch commented:
Thanks for the merge - and for the docs update (#20), completely missed it...

fajpunk commented:
Sure thing!  No worries about the docs.  4.0.0 is up on PyPI with this change.
